src/sync/impl/apple/network_reachability_observer.hpp

```diff
@@ -0,0 +1,63 @@
+////////////////////////////////////////////////////////////////////////////
+//
+// Copyright 2016 Realm Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+////////////////////////////////////////////////////////////////////////////
+
+#ifndef REALM_OS_NETWORK_REACHABILITY_OBSERVER_HPP
+#define REALM_OS_NETWORK_REACHABILITY_OBSERVER_HPP
+
+#include <functional>
+#include <string>
+
+#include <realm/util/cf_ptr.hpp>
+
+#include <SystemConfiguration/SystemConfiguration.h>
+
+namespace realm {
+
+enum NetworkReachabilityStatus {
+    NotReachable,
+    ReachableViaWiFi,
+    ReachableViaWWAN
+};
+
+class NetworkReachabilityObserver {
+public:
+    // An instanse to check if the default route is available.
```

Comment:
Monitoring of the specific host is too optimistic and in the current implementation is almost equal to the default route reachability because in case of the hostname reachability API generates the first notification for the default route and another notification after DNS query. I think having the ability to use this observer for both cases makes sense so it could be used on a client level as well as on the session level. Will make a unified constructor.